kwok/cloudprovider/helpers.go

```diff
@@ -30,78 +32,91 @@ import (
+		return nil, fmt.Errorf("could not parse JSON data: %w", err)
+	}
+
+	for _, opts := range instanceTypeOptions {
+		opts = setDefaultOptions(opts)
+		instanceTypes = append(instanceTypes, newInstanceType(opts))
+	}
+	return instanceTypes, nil
 }
 
-// InstanceTypesAssorted create many unique instance types with varying CPU/memory/architecture/OS/zone/capacity type.
-func ConstructInstanceTypes() []*cloudprovider.InstanceType {
-	var instanceTypes []*cloudprovider.InstanceType
-	for _, cpu := range []int{1, 2, 4, 8, 16, 32, 48, 64, 96, 128, 192, 256} {
-		for _, memFactor := range []int{2, 4, 8} {
-			for _, os := range []sets.Set[string]{sets.New(string(v1.Linux)), sets.New(string(v1.Windows))} {
-				for _, arch := range []string{v1beta1.ArchitectureAmd64, v1beta1.ArchitectureArm64} {
-					// Construct instance type details, then construct offerings.
-					mem := cpu * memFactor
-					pods := lo.Clamp(cpu*16, 0, 1024)
-					labels := MakeInstanceTypeLabels(cpu, memFactor)
-					opts := InstanceTypeOptions{
-						Name:             fmt.Sprintf("%s-%s-%s-%s", labels[InstanceFamilyLabelKey], labels[InstanceSizeLabelKey], arch, strings.Join(sets.List(os), ",")),
-						Architecture:     arch,
-						OperatingSystems: os,
-						Resources: v1.ResourceList{
-							v1.ResourceCPU:              resource.MustParse(fmt.Sprintf("%d", cpu)),
-							v1.ResourceMemory:           resource.MustParse(fmt.Sprintf("%dGi", mem)),
-							v1.ResourcePods:             resource.MustParse(fmt.Sprintf("%d", pods)),
-							v1.ResourceEphemeralStorage: resource.MustParse("20G"),
-						},
-						InstanceTypeLabels: labels,
-					}
-					price := PriceFromResources(opts.Resources)
-
-					opts.Offerings = cloudprovider.Offerings{}
-					for _, zone := range KwokZones {
-						for _, ct := range []string{v1beta1.CapacityTypeSpot, v1beta1.CapacityTypeOnDemand} {
-							opts.Offerings = append(opts.Offerings, cloudprovider.Offering{
-								CapacityType: ct,
-								Zone:         zone,
-								Price:        lo.Ternary(ct == v1beta1.CapacityTypeSpot, price*.7, price),
-								Available:    true,
-							})
-						}
-					}
-					instanceTypes = append(instanceTypes, newInstanceType(opts))
-				}
-			}
+func parseSizeFromType(ty, cpu string) string {
+	if matches := awsRegexp.FindStringSubmatch(ty); matches != nil {
+		return matches[1]
+	}
+
+	// If we can't figure out the size, fall back to the cpu value;
+	// this works for both Azure and GCP (and the "generic" instances
+	// generated by tools/gen_instances.go)
+	return cpu
+}
+
+func setDefaultOptions(opts InstanceTypeOptions) InstanceTypeOptions {
+	var cpu, memory string
+	for res, q := range opts.Resources {
+		switch res {
+		case "cpu":
+			cpu = q.String()
+		case "memory":
+			memory = q.String()
 		}
 	}
-	return instanceTypes
+	family := opts.Name[0:1] // all of the three major compute clouds (AWS, Azure, GCP) use the first letter of the instance name as the family
```

Comment:
In AWS, we call what you're talking about as category, where family encompasses the category + generation + other details. 

https://karpenter.sh/docs/concepts/scheduling/#:~:text=an%20instance%20category-,karpenter.k8s.aws/instance%2Dfamily,-g4dn

Is there a common string delimiter that we can use to infer what the family should be? If not, it's probably fine to just keep as is, and understand that we can't fully emulate all the CP instance types.

Reply:
So, AWS uses `(family).(size)`, GCP uses `(machine series)-(configuration)-(cpu count)`, and Azure (annoyingly) uses `[Family] + [Sub-family]* + [# of vCPUs] + [Constrained vCPUs]* + [Additive Features] + [Accelerator Type]* + [Version]`.  So we have things like

AWS: c5d.9xlarge
GCP: c3d-standard-4
Azure: M416ms_v2

I think I'm going to split on `[.-]` and if that fails, fall back to "first character", which will also work for the generic instance types you created.